ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/netty/NettyP2PNetwork.java

```diff
@@ -335,6 +339,32 @@ public void checkMaintainedConnectionPeers() {
     }
   }
 
+  @Override
+  public void attemptPeerConnections() {
+    // Start more connections until either we reach max peers or half our max peers quota is spent
+    // on pending connections. The limit on pending connections is to allow room for
+    // additional incoming connections
+    final int availablePeerSlots =
+        Math.min(maxPeers - connectionCount(), (maxPeers / 2) - pendingConnections.size());
+    if (availablePeerSlots > 0) {
+      final List<DiscoveryPeer> peers =
+          peerDiscoveryAgent
+              .getPeers()
+              .filter(peer -> peer.getEndpoint().getTcpPort().isPresent())
+              .filter(peer -> peer.getStatus() == PeerDiscoveryStatus.BONDED)
+              .filter(peer -> !isConnected(peer) && !isConnecting(peer))
+              .collect(Collectors.toList());
+      Collections.shuffle(peers);
+      LOG.info("Initiating connection to {} peers from the peer table", availablePeerSlots);
+      peers.stream().limit(availablePeerSlots).forEach(this::connect);
+      LOG.info(
+          "Connection count {} (pending: {}, connected: {})",
+          this.connectionCount(),
+          pendingConnections.size(),
+          connections.size());
+    }
+  }
```

Comment:
This is the key part of the change.  Pretty much all of the rest is just removing an unnecessary conversion from Stream to list when in every production case we then convert back to a stream anyway.